flask_restful/__init__.py

```diff
@@ -287,33 +297,86 @@ def handle_error(self, e):
             else:
                 raise e
 
-        headers = Headers()
-        if isinstance(e, HTTPException):
-            code = e.code
-            default_data = {
-                'message': getattr(e, 'description', http_status_message(code))
-            }
-            headers = e.get_response().headers
+        for typecheck, handler in self.errorhandlers:
+            if isinstance(e, typecheck):
+                try:
+                    # unpack(None) -> (None, _, _)
+                    data, code, headers = unpack(handler(e))
+                    if data:
+                        break
+                except:
+                    # log exception and return http 500
+                    current_app.log_exception(sys.exc_info())
+                    data, code, headers = self.default_http_exception_handler(
+                        InternalServerError('Exception in error handler!'))
+                    break
         else:
-            code = 500
-            default_data = {
-                'message': http_status_message(code),
-            }
+            # `default_exception_handler` is always at the bottom, even if no
+            # error handler was registered
+            data, code, headers = self.default_exception_handler(e)
+
+        return self.make_response(data, code, headers)
+
+    def default_exception_handler(self, e):
+        """Default response for an unhandled Exception
 
+        :param e: the exception raised while handling the request
+        :type e: Exception
+        :return: (data, code, headers) tuple
+        """
+        code = 500
+        default_data = { 'message': http_status_message(code), }
         data = getattr(e, 'data', default_data)
+        return data, code, Headers()
 
-        if code >= 500:
-            exc_info = sys.exc_info()
-            if exc_info[1] is None:
-                exc_info = None
-            current_app.log_exception(exc_info)
+    def default_http_exception_handler(self, e):
+        """Default response for a HTTPException
+
+        :param e: the exception raised while handling the request
+        :type e: werkzeug.exceptions.HTTPException
+        :return: (data, code, headers) tuple
+        """
+        code = e.code
+        default_data = { 'message': getattr(e, 'description', http_status_message(code)), }
+        data = getattr(e, 'data', default_data)
```

Comment:
`HTTPException` has no `data` property as I mentioned in [this comment](https://github.com/flask-restful/flask-restful/pull/544#issuecomment-189654298). I don't think we need to support this anymore and you can just return the `default_data`.